storage/replica.go

```diff
@@ -3177,6 +3177,12 @@ func (r *Replica) executeBatch(
 		optimizePuts(batch, ba.Requests, ba.Header.DistinctSpans)
 	}
 
+	// Update the node clock with the serviced request. This maintains a high
+	// water mark for all ops serviced, so that received ops without a timestamp
+	// specified are guaranteed one higher than any op already executed for
+	// overlapping keys.
+	r.store.Clock().Update(ba.Header.Timestamp)
```

Comment:
There's an awkward maybe-caveat here: if we update in a loop with each request, we pick up any forwarding that was made from the response of the previously executed one (or not? perhaps not). I don't think we've ever thought about this particularly hard, but we should now that we're trimming the fat. Can MVCC forward timestamps in any way that puts them ahead of the clock? I suppose it will on WriteTooOld, where it adds a logical tick.